### PR TITLE
Do not add duplicate presentation contexts

### PR DESCRIPTION
**Pull request by amoerie**
#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Do not add a presentation context if it is a complete, identical copy of an existing one

### CODE REVIEW

File: FO-DICOM.Core/Network/DicomPresentationContextCollection.cs [hunk 1]
```diff
@@ -79,6 +79,56 @@ public void Add(DicomUID abstractSyntax, params DicomTransferSyntax[] transferSy
         /// <param name="transferSyntaxes">Supported transfer syntaxes.</param>
         public void Add(DicomUID abstractSyntax, bool? userRole, bool? providerRole, params DicomTransferSyntax[] transferSyntaxes)
         {
+            transferSyntaxes = transferSyntaxes ?? Array.Empty<DicomTransferSyntax>();
+
+            // Double-check against duplicate presentation contexts
+            foreach (var existingPresentationContext in this)
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/Network/DicomPresentationContextCollection.cs]: Would it make sense to implement and use an equality comparer for this?

Comment (amoerie) [reply FO-DICOM.Core/Network/DicomPresentationContextCollection.cs]: I considered that but then you assign it a unique ID and it becomes unique that way. Unless that equality comparer would deliberately ignore the ID of course, but it seemed simpler like this. 
You'll find that there's also an overload that accepts a presentation context item and there is no check for duplicate values if you take that route. It is assumed you know what you're doing then.

Comment (mrbean-bremen) [reply FO-DICOM.Core/Network/DicomPresentationContextCollection.cs]: Ok, you are right - I wasn't sure about this, just thought it worth mentioning.

Comment (amoerie) [reply FO-DICOM.Core/Network/DicomPresentationContextCollection.cs]: Oh no it was a totally valid suggestion! 👍

**Comment by gofal** [comment]
This is a PR without an issue. So generally asked: what is the reason or issue for this change?
take the unittest as example. You are adding 5 presentation contexts to the association request. with this PR this is optimized to 4 presentation contexts. ok, thats fine. But the downside is now, that you have a n^2 runtime deep coparision check for each presentation context you add. so there are a lot of checks included now in every assocation for all clients in all circumstances, to reduce the number of sent bytes in some certain cases?

maybe some other solution, like an extra method "RemoveDuplicatePresentationContexts", that the user can call and that then checks all the presentation contexts before sending the request?

**Comment by gofal** [comment]
in case the idea of this PR would be, to solve the issue, that the maximum number of Presentation Contexts is exceeded but could be solved by reducing duplicate, then you could add a check before sending the AssocationRequest if the number is higher than the allowed. If this is true, then remove duplicates.

**Comment by amoerie** [comment]
> This is a PR without an issue. So generally asked: what is the reason or issue for this change? take the unittest as example. You are adding 5 presentation contexts to the association request. with this PR this is optimized to 4 presentation contexts. ok, thats fine. 

Sorry if this was missing a bit of context! I'm going through all our custom changes in our fork and - if they are not specific to our case - share them here.
This particular change was introduced because we have a configuration system related to transfer syntaxes per SOP Class UID, and it is possible to misconfigure it so that duplicate presentation contexts are the end result. Even though this is not necessarily a big problem, we did see PACS systems in the wild misbehave due to these duplicate presentation contexts.
On top of improving our configuration UI, we also added this safety net so that you can never have duplicate presentation contexts.

> But the downside is now, that you have a n^2 runtime deep coparision check for each presentation context you add. so there are a lot of checks included now in every assocation for all clients in all circumstances, to reduce the number of sent bytes in some certain cases?
> maybe some other solution, like an extra method "RemoveDuplicatePresentationContexts", that the user can call and that then checks all the presentation contexts before sending the request?

It's true that the worst case scenario now introduces a O(n²) algorithm, but this worst case scenario only occurs if it is always the same SOP Class UID with different transfer syntaxes. In reality, what we tend to see, is that the number of presentation contexts in an association and the number of transfer syntaxes per presentation context is quite limited. (<100 presentation contexts and < 5 transfer syntaxes per context).

Anyway, if you believe this feature is not worth the runtime cost, we can also always reject it. I personally don't see a lot of value in a dedicated `RemoveDuplicatePresentationContexts` method, because people will simply never call it. In that case, I'd rather drop the duplicate detection altogether.

**Comment by amoerie** [comment]
> But the downside is now, that you have a n^2 runtime deep coparision check for each presentation context you add. so there are a lot of checks included now in every assocation for all clients in all circumstances, to reduce the number of sent bytes in some certain cases?
> maybe some other solution, like an extra method "RemoveDuplicatePresentationContexts", that the user can call and that then checks all the presentation contexts before sending the request?

To address your performance concerns (which were valid, even though I still believe the real world impact would have been small), I have switched the implementation to a HashSet based one.

We now keep a hash set of existing presentation contexts, and only add a new presentation context if there is no collision with the hash set. Instead of a n^2 loop when adding a new presentation context, we now calculate the hash code of this presentation context and look it up in the hashset, which should be a O(1) operation.
We do consume a little bit of extra memory for the hash set, but this collection only contains references to the presentation contexts, so the overhead should be extremely tiny.

File: FO-DICOM.Core/Network/DicomPresentationContextCollection.cs [hunk 2]
```diff
@@ -96,7 +104,11 @@ public void Add(DicomUID abstractSyntax, bool? userRole, bool? providerRole, par
         /// <exception cref="T:System.NotSupportedException">The <see cref="T:System.Collections.Generic.ICollection`1"/> is read-only.</exception>
         public void Add(DicomPresentationContext item)
         {
-            _pc.Add(item.ID, item);
+            // Double-check against duplicate presentation contexts
+            if (_uniquePcs.Add(item))
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/Network/DicomPresentationContextCollection.cs:108]: The other possibility would be to remove an existing PC and add it to the end, which would change the order - but I think this is not something the user would think about, so it is most probably ok as it is.

**Review by mrbean-bremen — APPROVED** [review]

File: FO-DICOM.Core/Network/DicomPresentationContextCollection.cs (hunk 2 above)
Comment (amoerie) [reply FO-DICOM.Core/Network/DicomPresentationContextCollection.cs:108]: If the user wants to add it at the end, he can always the existing one first. I think the current implementation makes the most sense now..

**Pull request merged**